### DISCUSSION TITLE
Product Collection Layout Control Changes

### DISCUSSION THREAD

**Discussion by ObliviousHarmony** [woocommerce/woocommerce#47437] [WooCommerce, Blocks, and Site Editing]
## Background

In https://github.com/WordPress/gutenberg/pull/49050 Gutenberg's Query Loop block migrated from the custom `displayLayout` attribute to `supports.layout` to support block spacing. The `layout` feature has a number of different display modes and utilizing this allows the Query Loop block to automatically benefit from future improvements. The `grid` type, for example, powers the Grid block and all of the neat features that come with it.

Someone [requested that the Product Collection block also support block spacing](https://github.com/woocommerce/woocommerce/issues/43307), and as a result, I took a look at making the same change as Gutenberg in https://github.com/woocommerce/woocommerce/pull/46255. I discovered that `supports.layout` does not currently provide the same kind of responsiveness that the Product Collection block does and as a result we need to make a decision about how to move forward.

## Design Change

The `grid` layout type only supports two (mutually exclusive) configuration options:

* **Column Count**: Sets a static number of columns regardless of the container's width. This is _not_ responsive.
* **Minimum Column Width**: Sets the minimum width for each column. This makes the column count responsive, however, _there is no way to set a maximum column count_.

While there _is_ support for responsiveness, the Product Collection block currently uses a maximum column count with a hard-coded minimum width of 150px. I do, however, want to highlight that there is [a draft PR](https://github.com/WordPress/gutenberg/pull/60941) open that adds support for setting a maximum column count. While this means in the future we would have support for this automatically re-introduced, for now, we would be removing that feature from the Product Collection block.

As an added bonus, by using `supports.layout`, all of the built-in supported layout theme features would work with the Product Collection block!

<img width="277" alt="Screenshot 2024-05-13 at 14 56 35" src="https://github.com/woocommerce/woocommerce/assets/6451942/ef2297ef-2a53-4965-b227-770c6d5c7354">
<img width="277" alt="Screenshot 2024-05-13 at 14 56 42" src="https://github.com/woocommerce/woocommerce/assets/6451942/77ad1e72-2591-4c1c-879a-ed02929c40e1">

## Alternatives

Before continuing I would like to highlight the alternatives I've looked at:

### Custom Block Spacing Support

One thing worth noting is that supporting block spacing means we _can't_ use CSS classes for column divisions anymore. A caveat to CSS Grid is that the `gap` property is not excluded when setting the width. You need to explicitly subtract it from the column width yourself. As a consequence we need to generate runtime styles for the template block. **We should also be thinking about theme support and add any necessary options for that as well**.

At that point we are essentially just reinventing the wheel and doing the same work that `supports.layout` does. This seems heavy-handed just to have this one small difference. I haven't dove _too_ far down this route, however, there also looks like there might be problems with inner block layouts which can inherit some options from their parents. We _could_ see this as a polyfill but this is the only block we would use it with.

### Custom Style Overrides

We _could_ inject our own styles to override the column count. This would let us use `supports.layout` while still having the same style of responsiveness. The danger here though is that it means future changes to `layout` might break our custom styles. Personally, I think using `supports.layout` is an all-or-nothing situation.

## Requested Feedback

In the interest of moving this forward, the only _real_ point of contention is around the change in responsiveness. The goal of a "responsive" Product Collection is to have the number of columns change based on the size of the container. In my opinion, it doesn't really matter whether or not there is a maximum that you can set. With that said, it _does_ look like this would be a feature coming in the future from Gutenberg (ergo, a future WordPress version).

What do you folks think?

**Comment by manospsyx** [comment]
We have previously dealt with this problem in Composite Products and Product Bundles, where product collections are often embedded in wrappers that are themselves nested in columns. 

We know from end-users, site builders and theme creators that they need enough control to:

- Set a **minimum** number of grid columns.
- Set a **maximum** number of columns. 

When users make this request, there is an important piece of information that is implied: 

In many scenarios, a solution based on screen breakpoints will fail, because it's the container of our grid that provides the constraints; and its width cannot be known. Minimum and maximum column sizes are only meaningful relative to some container width thresholds (top and bottom).

---

To directly serve the user/developer needs we've identified, I believe that the most intuitive approach would be to surface two options

- "Min Columns" and 
-  "Max Columns"

....which users understand intuitively.

For example:

- Below some pre-defined, hardcoded, but filterable **container** width, visitors will see the specified "Min Columns".
- Above some pre-defined, hardcoded, but filterable **container** width, visitors will see the specified "Max Columns" ("max above a threshold").
- Between these two thresholds, the number of columns will vary from the specified min number to the specified max number (requires a calculation of column width?). 


> The grid layout type only supports two (mutually exclusive) configuration options

- **Manual** mode sounds like it's meant to be unresponsive, and therefore mobile-unfriendly. 
- **Auto** mode combined with the **Min Column Width** provides some flexibility. However:
  - **Min Column Width** comes with friction, because it doesn't answer the need directly (set column counts).

**All that said...**

> Personally, I think using supports.layout is an all-or-nothing situation.

I agree with this. Rolling out a custom solution for Woo does not seem tenable to me. What's crucial here is that we help Gutenberg developers better understand user needs, and help them implement a more flexible solution that doesn't require theme developers and site builders to write custom CSS -- or maintain it. 

Until we solve the problem at the block level, and for any type of grid (not just products!) the burden of it will be shifted to developers and builders.

**Reply by ObliviousHarmony** [reply]
> Manual mode sounds like it's meant to be unresponsive, and therefore mobile-unfriendly.

Absolutely, and imo, I don't actually feel like manual mode really serves a purpose beyond "existing" for the sake of it.

> Min Column Width comes with friction, because it doesn't answer the need directly (set column counts).

As noted though, there _is_ a draft PR adding that feature so this wouldn't be a permanent limitation. Given our L-2 WordPress compatibility policy, even if it was added to Gutenberg ahead of WordPress 6.6, we wouldn't be able to acknowledge it as the "default" until next year. (We can still utilize it we'd just be in a middle-ground where it didn't work in all WordPress versions we support.)

> I agree with this. Rolling out a custom solution for Woo does not seem tenable to me.

With that in mind, how would you proceed moving forward? Should we just not support block spacing as a configuration option until we can support the same responsive columns that we have now?

**Reply by sunyatasattva** [reply]
> In many scenarios, a solution based on screen breakpoints will fail, because it's the container of our grid that provides the constraints; and its width cannot be known. 

[A man can dream…](https://caniuse.com/css-container-queries)

> pre-defined, hardcoded, but filterable container width

What do you mean with “predefined, hardcoded, _but filterable_”?

>> @manospsyx: I agree with this. Rolling out a custom solution for Woo does not seem tenable to me. What's crucial here is that we help Gutenberg developers better understand user needs, and help them implement a more flexible solution that doesn't require theme developers and site builders to write custom CSS -- or maintain it.

> @ObliviousHarmony:With that in mind, how would you proceed moving forward? Should we just not support block spacing as a configuration option until we can support the same responsive columns that we have now?

I might be misunderstanding here, but I'm not too sure I entirely agree with that. Given the recent proposal by @joshuatf (pcShBQ-1LM-p2), I think it is acceptable for us to move forward with having the cake and eat it too. Like @ObliviousHarmony has said, we are bound by the L-2 compatibility policy, and we might feel like we have to decide whether we want the block spacing (and all the other nice things that come with it) and renounce to responsiveness, or the responsiveness and fall behind in Gutenberg standards.

My suggestion here is that we follow the process proposed by @joshuatf: we align as much as possible with Gutenberg and core folks on how the feature will look like (given that there seems to be interest in adding such a feature), and duplicate the code in Woo in a compatibility layer that can be removed when the support policy expires.

In this specific case, it would be also interesting to see if we can somehow accelerate the process: I suppose there is _some kind_ of public API (e.g. the attribute ids, I think?), but I wonder if we necessarily have to wait for merge, or can just get away with waiting for consensus on those.

Don't have a strong opinion about that, but my 2c are just that this doesn't have to be an _either/or_ kind of situation.

**Reply by kmanijak** [reply]
I'm with @sunyatasattva on this one. At least in theory we should be able to finalise the migration and have a missing feature (max number of columns). Obviously, the prerequisite would be to agree on and finalise the upstream contribution so we can use exactly same API agreed upstream (for easier switch from "local" implementation to Gutenberg's one in a year or so).

In practice though, I'm not exactly sure how hard (or easy) it is to extend the Gutenberg's layout support in our repo, so it does provide the required feature. Would we use a `patch-package`? Worth exploring it.

---

But in any case I see a sentiment from @manospsyx that the ability to set max number of columns is a must (and I tend to agree!). Having said that, I think we should explore option provided by @sunyatasattva. In ideal scenario we don't introduce any regression and that means I vote for this issue _not to be_ included in "Remove Beta Label" epic.

**Reply by manospsyx** [reply]
> What do you mean with “predefined, hardcoded, but filterable”?

= defined in code, not exposed as a UI option, but override-able by developers.

> [A man can dream…](https://caniuse.com/css-container-queries)

wow... cc @xristos3490

**Comment by nerrad** [comment]
> I'm with @sunyatasattva on this one. At least in theory we should be able to finalise the migration and have a missing feature (max number of columns). Obviously, the prerequisite would be to agree on and finalise the upstream contribution so we can use exactly same API agreed upstream (for easier switch from "local" implementation to Gutenberg's one in a year or so).

Yes. Work towards this for sure.

> One thing worth noting is that supporting block spacing means we can't use CSS classes for column divisions anymore.

Will this impact stores that have already implemented custom styling targeting these CSS classes? If so, it's not necessarily a blocker but does need to be accounted for.
